test/path_tokenizer.h

```diff
@@ -0,0 +1,56 @@
+//
+// Created by wkl04 on 12-12-2019.
+//
+
+#ifndef MQTTSUBSCRIPTION_PATH_TOKENIZER_H
+#define MQTTSUBSCRIPTION_PATH_TOKENIZER_H
+
+#include <algorithm>
+
+template<typename Output>
+inline void mqtt_path_tokenizer(MQTT_NS::string_view str, Output write)
+{
+    auto first = str.begin();
+    auto last = str.end();
+
+    while (true) {
+        auto pos = std::find(first, last, '/');
+        if(!write(MQTT_NS::string_view(first, std::distance(first, pos)))) {
+            break;
+        }
+
+        if (pos == last) {
+            break;
+        }
+
+        first = std::next(pos);
+    }
+}
+
+// See if the topic is valid (does not contain # or +)
+inline bool mqtt_valid_topic(MQTT_NS::string_view topic)
+{
+    if(topic.empty()) {
+        return false;
+    }
+
+    if(topic[0] == '$') {
+        return false;
+    }
+
+    return (topic.find_first_of("#+") == std::string::npos);
+}
+
+// See if the subscription is valid
+inline bool mqtt_valid_subscription(MQTT_NS::string_view subscription)
```

Comment:
Is this really sufficient?

I check subscriptions this way:

```cpp

inline bool validate_topic_pattern(std::string_view topicPattern)
{
    /*
     * Confirm the topic pattern is valid before registering it.
     * Use rules from http://docs.oasis-open.org/mqtt/mqtt/v3.1.1/os/mqtt-v3.1.1-os.html#_Toc398718106
     */
    for(size_t idx = topicPattern.find_first_of("+#");
        std::string_view::npos != idx;
        idx = topicPattern.find_first_of("+#", idx+1))
    {
        BOOST_ASSERT(   ('+' == topicPattern[idx])
                     || ('#' == topicPattern[idx]));
        if('+' == topicPattern[idx])
        {
            /*
             * Either must be the first character,
             * or be preceeded by a topic seperator.
             */
            if(   (0 != idx)
               && ('/' != topicPattern[idx-1]))
            {
                return false;
            }
            /*
             * Either must be the last character,
             * or be followed by a topic seperator.
             */
            if(   (topicPattern.size()-1 != idx)
               && ('/' != topicPattern[idx+1]))
            {
                return false;
            }
        }
        else // multilevel wildcard
        {
            /*
             * Must be absolute last character.
             * Must only be one multi level wild card.
             */
            if(idx != topicPattern.size()-1)
            {
                return false;
            }
            /*
             * If not the first character, then the
             * immediately preceeding character must
             * be a topic level separator.
             */
            if(   (0 != idx)
               && ('/' != topicPattern[idx-1]))
            {
                return false;
            }
        }
    }
    return true;
}
```


You'll note that I don't check for $, and I don't validate shared subscriptions,  which I need to do eventually.

Reply:
Indeed a '+' should be followed by a '/', which I do not check. Maybe we should use some regex checker ? 

The $ is actually valid in topic and subscriptions, i think it can also be anywhere in the topic, only as the first character $SYS/ .. it marks a special / shared subscription

Reply:
An empty topic is valid or not ?

Reply:
Your validate function is already part of mqtt_cpp or not ?

Reply:
I believe that an empty topic is not valid, but it is very late in the night for me, so I am not thinking at full power, so I could be wrong.

This function is not part of mqtt_cpp, yet. It is also not complete.

I would be happy to see this function included into mqtt_cpp, though. So you're welcome to include it (or an improved version) in your pull request.